environment/environment.go

```diff
@@ -227,42 +225,67 @@ func (env *Environment) UpdateConfig(ctx context.Context, explanation string, ne
 
 	env.Config = newConfig
 
-	// Re-build the base image from the worktree
-	container, err := env.buildBase(ctx)
+	// Get current working directory from container to preserve changes
+	currentWorkdir := env.container().Directory(env.Config.Workdir)
+
+	// Re-build the base image with the new config
+	container, err := env.buildBase(ctx, currentWorkdir)
 	if err != nil {
 		return err
 	}
 
-	if err := env.apply(ctx, "Update environment", explanation, "", container); err != nil {
+	if err := env.apply(ctx, container); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (env *Environment) Run(ctx context.Context, explanation, command, shell string, useEntrypoint bool) (string, error) {
+func (env *Environment) Run(ctx context.Context, command, shell string, useEntrypoint bool) (string, error) {
 	args := []string{}
 	if command != "" {
 		args = []string{shell, "-c", command}
 	}
 	newState := env.container().WithExec(args, dagger.ContainerWithExecOpts{
 		UseEntrypoint: useEntrypoint,
+		Expect:        dagger.ReturnTypeAny, // Don't treat non-zero exit as error
```

Comment:
note: I _think_ this can have weird caching side effects by caching failed operations.

Particularly annoying for transient failures that won't get busted by content (e.g. `pip install foo` --> `connection timeout` --> will always return `connection timeout` from cache without trying).

Although not 100% sure the engine doesn't have workarounds already in place for that (/cc @sipsma ?)

Reply:
As in -- on an operation-by-operation level it's all simple -- we manually check `ExitCode` and return accordingly.

However, when resuming from the `Container` state blob,  we can't manually check ExitCodes for each exec in the chain and are relying on 1) the cache to not re-run execs OR 2) if it's not cached, the engine will re-exec and bubble up exec errors along the way, if any

I think `dagger.ReturnTypeAny` will prevent that, though. Emphasis on *think*.

Reply:
in my mental model, yes, this gets kinda broken in the case of new failures on cache bust.